sdk/tests/iot/hub/test_az_iot_hub_client_properties.c

```diff
@@ -0,0 +1,1886 @@
+  az_iot_hub_client_options options = az_iot_hub_client_options_default();
+  options.component_names = test_components;
+  options.component_names_length = test_components_length;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, &options), AZ_OK);
+
+  az_iot_hub_client_properties_response_type response_type
+      = AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_GET;
+  az_json_reader jr;
+  int32_t version;
+  assert_int_equal(az_json_reader_init(&jr, test_property_payload_long, NULL), AZ_OK);
+
+  assert_int_equal(
+      az_iot_hub_client_properties_get_properties_version(&client, &jr, response_type, &version),
+      AZ_OK);
+  assert_int_equal(version, 30);
+}
+
+static void test_az_iot_hub_client_properties_get_properties_version_out_of_order_succeed()
+{
+  az_iot_hub_client client;
+  az_iot_hub_client_options options = az_iot_hub_client_options_default();
+  options.component_names = test_components;
+  options.component_names_length = test_components_length;
+  assert_int_equal(
+      az_iot_hub_client_init(&client, test_device_hostname, test_device_id, &options), AZ_OK);
+
+  az_iot_hub_client_properties_response_type response_type
+      = AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_GET;
+  az_json_reader jr;
+  int32_t version;
+  assert_int_equal(az_json_reader_init(&jr, test_property_payload_out_of_order, NULL), AZ_OK);
+
+  assert_int_equal(
+      az_iot_hub_client_properties_get_properties_version(&client, &jr, response_type, &version),
+      AZ_OK);
+  assert_int_equal(version, 4);
+}
+
+// test_get_next_component_property invokes az_iot_hub_client_properties_get_next_component_property
+// and verifies that the component and property match expected.  It then automatically advances
+// the json_reader.
+static void test_get_next_component_property(
+    const test_az_properties_expected* expected,
+    az_iot_hub_client_properties_response_type response_type,
+    az_iot_hub_client_property_type property_type,
+    az_iot_hub_client* client,
+    az_json_reader* jr,
+    az_span* component_name)
+{
+  assert_int_equal(
+      az_iot_hub_client_properties_get_next_component_property(
+          client, jr, response_type, property_type, component_name),
+      AZ_OK);
+
+  if (az_span_is_content_equal((az_span)TEST_PROPERTY_EMPTY_COMPONENT, expected->component_name))
+  {
+    assert_true(az_span_is_content_equal(AZ_SPAN_EMPTY, *component_name));
+  }
+  else
+  {
+    assert_true(az_span_is_content_equal(*component_name, expected->component_name));
+  }
+
+  assert_true(az_json_token_is_text_equal(&jr->token, expected->property_name));
+  assert_int_equal(az_json_reader_next_token(jr), AZ_OK);
+  assert_int_equal(jr->token.kind, expected->token_kind);
+
+  if (jr->token.kind == AZ_JSON_TOKEN_NUMBER)
+  {
+    double d;
+    assert_int_equal(az_json_token_get_double(&jr->token, &d), AZ_OK);
+    assert_true(fabs(d - expected->u.number) < test_max_allowed_error);
```

Comment:
`test_max_allowed_error`: I saw the value for this and am curious what it's doing?

Reply:
the number comes back to us as a double, so I can't just do a `==` on it because of floating point precision.  

Maybe `test_max_allowed_double_tolerance` or something, since this isn't really an error.

Reply:
ahh yes tolerance would be better. I was thinking in the context of some maximum return value or error